core/include/detray/core/detail/indexing.hpp

```diff
@@ -37,44 +39,307 @@ DETRAY_HOST inline std::ostream& operator<<(std::ostream& os,
+                                                   static_cast<value_t>(u));
+        }
+        return *this;
+    }
+
+    /// Shift both indices by @param s. Allow negative shift
+    DETRAY_HOST_DEVICE
+    constexpr index_range& shift(long s) {
+        assert(static_cast<long>(lower()) + s >= 0l);
+
+        // Update lower index
+        encoder::template set_bits<lower_mask>(
+            m_value, static_cast<value_t>(static_cast<long>(lower()) + s));
+        // Shift upper index, if range contains upper index instead of size
+        if constexpr (!contains_size) {
+            encoder::template set_bits<upper_mask>(
+                m_value, static_cast<value_t>(static_cast<long>(upper()) + s));
+        }
+
+        return *this;
+    }
+
+    /// Check wether the index range is valid to use.
+    DETRAY_HOST_DEVICE
+    constexpr bool is_invalid() const noexcept {
+        return encoder::template is_invalid<lower_mask, upper_mask>(m_value);
+    }
+
+    /// Comparison operators
+    /// @{
+    /// Equality operator
+    constexpr bool operator==(const index_range& other) const {
+        return (lower() == other.lower()) && (upper() == other.upper());
+    };
+
+    /// Comparison operators
+    DETRAY_HOST_DEVICE
+    friend constexpr auto operator<=>(const index_range lhs,
+                                      const index_range rhs) noexcept {
+        const auto l{lhs.size()};
+        const auto r{rhs.size()};
+        if (l < r || (l == r && l < r)) {
+            return std::strong_ordering::less;
+        }
+        if (l > r || (l == r && l > r)) {
+            return std::strong_ordering::greater;
+        }
+        return std::strong_ordering::equivalent;
+    }
+    /// @}
+
+    /// Arithmetic operators
+    /// @{
+    /// @returns extended index range that contains both initial ranges
+    DETRAY_HOST_DEVICE
+    friend index_range operator+(const index_range lhs, const index_range rhs) {
+        const index_t new_lower{lhs.lower() < rhs.lower() ? lhs.lower()
+                                                          : rhs.lower()};
+        const index_t new_upper{lhs.upper() > rhs.upper() ? lhs.upper()
+                                                          : rhs.upper()};
+
+        return new_lower < new_upper
+                   ? index_range{}.set_lower(new_lower).set_upper(new_upper)
+                   : index_range{}.set_lower(new_upper).set_upper(new_lower);
+    }
+
+    /// Expands upper range boundary by @param index
+    DETRAY_HOST_DEVICE
+    friend index_range operator+(const index_range lhs,
+                                 const index_type index) {
+        return index_range{}
+            .set_lower(lhs.lower())
+            .set_upper(lhs.upper() + index);
+    }
+
+    /// @returns index range that was reduced to minimal coverage
+    DETRAY_HOST_DEVICE
+    friend index_range operator-(const index_range lhs, const index_range rhs) {
```

Comment:
From my experience, operator overloads are usually a bad idea for this. 

From the call site, it's impossible to tell what's happening.

Reply:
I think the main reason for this was, that the index range and a single numeric index are used interchangeably in some cases (some detectors do not use multiple masks per surface). In the end, I think I did end up special casing all of this, but never removed the operators...

Reply:
OK, so could be removed then in future.